kickstarter/ios-oss
# 💲[Native Checkout] Sheet Overlay Container

# 📲 What

This PR adds `SheetOverlayContainerViewController`, a view controller that's intended to act as a custom container to **any** other view controller, rendering the child VC in a "card-like" way. It also comes with it's own custom transition animator, `SheetOverlayTransitionAnimator` which performs a custom modal presentation transition.

What the `SheetOverlay` does:
- contains any view controller within a "card" view that displays over the presenting view controller
- animates the presentation in a custom way

What the `SheetOverlay` **does not** do:
- manage navigation for you (dismissing, pushing, popping VCs)
- pass messages to/from the contained VC

Also in this PR:
- refactored the `Back this project` button's container view's layer to use `maskedCorners` to mask only the top right and top left corners, instead of creating a custom shape layer to achieve this. `maskedCorners` is only available on iOS 11 and above, which is fine given we're in the process of deprecating iOS 10

# 🤔 Why

This is required to present existing view controllers with the "card" design, as required in several places during the new checkout flow.

# 🛠 How

**New**
- `SheetOverlayContainerViewController`
- `SheetOverlayTransitionAnimator`


# 👀 See

**SheetOverlayContainerViewController configured with LoginToutViewController**
![Simulator Screen Shot - iPhone 8 - 2019-05-09 at 11 54 10](https://user-images.githubusercontent.com/3156796/57473131-23396600-725d-11e9-8f75-6bb7fe7ea4cf.png)
![Simulator Screen Shot - iPhone 8 - 2019-05-09 at 11 54 14](https://user-images.githubusercontent.com/3156796/57473140-26cced00-725d-11e9-9992-41ee8a124bc2.png)

![vKzZoX1mTv](https://user-images.githubusercontent.com/3156796/57473800-ac9d6800-725e-11e9-99ef-8b4bc53443ac.gif)


**SheetOverlayContainerViewController configured with AddNewCardViewController**
![Simulator Screen Shot - iPhone X - 2019-05-09 at 11 57 26](https://user-images.githubusercontent.com/3156796/57473162-377d6300-725d-11e9-8822-d6796b34cdb4.png)


# ♿️ Accessibility 

- [x] Tap targets use minimum of 44x44 pts dimensions
- [x] Works with VoiceOver
- [x] Supports Dynamic Type 

# 🏎 Performance

- [ ] Optimized Blended Layers (screenshots)

# ✅ Acceptance criteria

**How to test**

In `ProjectNavigatorViewController`, paste the following code into the function on line 171:

```
    let loginOutVC = LoginToutViewController.configuredWith(loginIntent: .backProject)
    let navigationController = UINavigationController(rootViewController: loginOutVC)
    let containerVC = SheetOverlayContainerViewController(childViewController: navigationController,
                                                          childViewOffset: 50)

    self.present(containerVC, animated: true)
```

Note that you can initialize `SheetOverlayContainerViewController` with *any* view controller.

- [ ] Log in on staging with an admin user. Navigate to any project. Tap "Back this project". The login/signup screen should animate from the bottom with a "sheet" animation, and partially cover the underlying screen.
